quickwit/quickwit-search/src/leaf.rs

```diff
@@ -414,6 +415,111 @@ pub(crate) fn rewrite_start_end_time_bounds(
 }
 
+#[derive(Debug)]
+enum MutCanSplitDoBetter {
+    Uninformative,
+    SplitIdHigher(Option<String>),
+    SplitTimestampHigher(Option<i64>),
+    SplitTimestampLower(Option<i64>),
+}
+
+#[derive(Debug, Clone)]
+pub struct CanSplitDoBetter {
+    inner: Arc<RwLock<MutCanSplitDoBetter>>,
+}
+
+impl CanSplitDoBetter {
+    /// Split metadata contains no information using for sorting
+    pub fn uninformative() -> Self {
+        CanSplitDoBetter {
+            inner: Arc::new(RwLock::new(MutCanSplitDoBetter::Uninformative)),
+        }
+    }
+
+    /// No order was requested, split id can be useful
+    pub fn no_order() -> Self {
+        CanSplitDoBetter {
+            inner: Arc::new(RwLock::new(MutCanSplitDoBetter::SplitIdHigher(None))),
+        }
+    }
+
+    /// Sorted by desc timestamp was requested
+    pub fn desc_timestamp() -> Self {
+        CanSplitDoBetter {
+            inner: Arc::new(RwLock::new(MutCanSplitDoBetter::SplitTimestampHigher(None))),
+        }
+    }
+
+    /// Sorted by asc timestamp was requested
+    pub fn asc_timestamp() -> Self {
+        CanSplitDoBetter {
+            inner: Arc::new(RwLock::new(MutCanSplitDoBetter::SplitTimestampLower(None))),
+        }
+    }
+
+    /// Returns whether the given split can possibly give documents better than the one already
+    /// known to match.
+    pub fn can_be_better(&self, split: &SplitIdAndFooterOffsets) -> bool {
+        match &*self.inner.read().unwrap() {
+            MutCanSplitDoBetter::SplitIdHigher(Some(split_id)) => split.split_id >= *split_id,
+            MutCanSplitDoBetter::SplitTimestampHigher(Some(timestamp)) => {
+                split.timestamp_end() > *timestamp
+            }
+            MutCanSplitDoBetter::SplitTimestampLower(Some(timestamp)) => {
+                split.timestamp_start() < *timestamp
+            }
+            _ => true,
+        }
+    }
+
+    /// Record the new worst-of-the-top document, that is, the document which would first be
+    /// evicted from the list of best documents, if a better document was found. Only call this
+    /// funciton if you have at least max_hits documents already.
+    pub fn record_new_worst_hit(&self, hit: &PartialHit) {
+        match &mut *self.inner.write().unwrap() {
+            MutCanSplitDoBetter::Uninformative => (),
+            MutCanSplitDoBetter::SplitIdHigher(split_id) => *split_id = Some(hit.split_id.clone()),
+            MutCanSplitDoBetter::SplitTimestampHigher(timestamp) => {
+                if let Some(SortValue::I64(timestamp_ns)) = hit.sort_value {
+                    // if we get a timestamp of, says 1.5s, we need to check up to 2s to make
+                    // sure we don't throw away something like 1.2s, so we should round up while
+                    // dividing.
+                    let mut timestamp_s = timestamp_ns / 1_000_000_000;
+                    if timestamp_ns % 1_000_000_000 != 0 {
+                        timestamp_s += 1;
+                    }
```

Comment:
I think the usual way to compute ceil_div is the following
```
#[inline]
fn ceil_div(n: i64, q: i64) {    
    assert!(q != 0i64);
    assert!(q != i64::MAX);
    (n + (q  - 1)) / q
}
```

```suggestion
      ceil_div(timestamp_ns, 1_000_000)
```